### PR TITLE
[Experimental] Signing write requests

### PR DESCRIPTION
**Pull request by alanprot**
**What this PR does**:
Sign the write requests to prevent future data corruption as https://github.com/cortexproject/cortex/issues/5419  

**Which issue(s) this PR fixes**:
Related to: #5419

**Checklist**
- [X] Tests updated
- [X] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/grpcclient/grpcclient.go [hunk 1]
```diff
@@ -29,8 +29,9 @@ type Config struct {
 	BackoffOnRatelimits bool           `yaml:"backoff_on_ratelimits"`
 	BackoffConfig       backoff.Config `yaml:"backoff_config"`
 
-	TLSEnabled bool             `yaml:"tls_enabled"`
-	TLS        tls.ClientConfig `yaml:",inline"`
+	TLSEnabled         bool             `yaml:"tls_enabled"`
+	TLS                tls.ClientConfig `yaml:",inline"`
+	EnableSignRequests bool             `yaml:"-"`
```
Comment (alvinlin123) [inline pkg/util/grpcclient/grpcclient.go:34]: nit: for consistency I think `SignRequestsEnabled` is a more consistent naming convention as above :)

File: pkg/util/grpcclient/signing_handler.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	return handler(ctx, req)
+}
+
+func UnarySigningClientInterceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	rs, ok := req.(SignRequest)
+
+	if !ok {
+		return invoker(ctx, method, req, reply, cc, opts...)
```
Comment (alvinlin123) [inline pkg/util/grpcclient/signing_handler.go:64]: Shouldn't we return error in this case? Or we are worry about "intermediate state" where both signed and unsigned request may exist? If this is the case, I wonder if we should emit a metrics to monitor if all requests are signed etc.

File: pkg/util/grpcclient/signing_handler.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	md, ok := metadata.FromOutgoingContext(ctx)
+	if !ok {
+		md = metadata.New(map[string]string{})
+	}
+
+	newCtx := ctx
+
+	if s, ok := md[reqSignHeaderName]; ok {
+		if len(s) == 1 {
+			if s[0] != signature {
```
Comment (alvinlin123) [inline pkg/util/grpcclient/signing_handler.go:82]: nit: merged this `if` with the above if statement will save you on indentation, which makes the code look nicer :) 
like `if len(s) == 1 && s[0] != signature`  

or maybe just one gigantic if ... 

```
if len(s) != 1 || s[0] != signagure {
   return ErrDifferentSignaturePresent
}
```

File: pkg/util/grpcclient/signing_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/user"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+)
+
+func TestUnarySigningHandler(t *testing.T) {
```
Comment (alvinlin123) [inline pkg/util/grpcclient/signing_handler_test.go:15]: Should we have more test cases for testing `ErrDifferentSignaturePresent` error condition?

**Review by alvinlin123 — APPROVED** [review]

File: pkg/util/grpcclient/signing_handler.go (hunk 2 above)
Comment (alanprot) [reply pkg/util/grpcclient/signing_handler.go:64]: This handlers wrapper all Unary calls so many other types of requests does not implement this interface.

The behavior is secured by a unit test though.

File: pkg/util/grpcclient/signing_handler_test.go (hunk 4 above)
Comment (alanprot) [reply pkg/util/grpcclient/signing_handler_test.go:15]: i can but i would need to manually force that error. lemme do that.

File: pkg/util/grpcclient/signing_handler.go (hunk 3 above)
Comment (alanprot) [reply pkg/util/grpcclient/signing_handler.go:82]: Then i need to check again on the else (len > 1), right?

Comment (alvinlin123) [reply pkg/util/grpcclient/signing_handler.go:83]: Oh i'm blind I thought you are returning `ErrDifferentSignaturePresent` in the case of len > 1 too. 

My bad.

**Review by yeya24 — APPROVED** [review]

**Review by friedrichg — APPROVED** [review]

**Pull request merged**